### DISCUSSION TITLE
P2P communication for ranging (+ logging)

### DISCUSSION THREAD

**Discussion by shossains** [bitcraze/discussions#1138] [Q&A] [answered]
Hi there,

I am currently working with the P2P API to send messages back and forth between 2 Crazyflies.
My goal is to send these messages in the order of tens of nanosecond (100 MHz).
With the travel duration of the packets I will able to determine the distance of the drones from each other.

From the example [app_peer_to_peer.c](https://github.com/bitcraze/crazyflie-firmware/blob/master/examples/app_peer_to_peer/src/peer_to_peer.c) there was a [vTaskDelay](https://github.com/bitcraze/crazyflie-firmware/blob/101de77a6746e99410074bd79da6f02e33f4e6c6/examples/app_peer_to_peer/src/peer_to_peer.c#L93) present to prevent collisions of the messages.
In my implementation I am sending packets constantly (using a infinite loop). Removing the task delay makes the drones not properly booting up, and adding a delay as little as 1 tick makes it dependent on the tick frequency of FreeRTOS (which seems to be capped at 1000Hz).

**Is there a way to send these packets P2P at a rate of 100MHz (or at least higher than the tick frequency of FreeRTOS)?**

And another question is how does logging work exactly? Is it always at a fixed rate of 10,000Hz? **Can I log variables only when they are updated?**

TYSM in advance (:

**Comment by ataffanel** [comment, marked as answer]
This is going to be impossible using the Crazyflie stock hardware:

While the STM32 is running at slightly more than 100MHz, it only runs at 168MHz. So if you wanted to do something 100M times per seconds you would have 1.68 instructions to do so. Otherwise the way to have code executing faster than the OS tick is to setup a hardware timer and interrupt by yourself.

The case is even worst on the radio nRF51822 side where the CPU runs at 16MHz and the radio communication at 2MBit/s, so at 100MHz you cannot event send one bit of radio. Things are made even worse by the fact that while the nRF51822 can time packet with a timer, the timer can only run at 1MHz.

Some people have tried to range with the 2.4GHz radio of the Crazyflie, but this was done by sending a lot of packet and trying to dither the 1MHz timer output to increase its precision over a lot of measurement. The challenge of course when ranging with 2.4GHz is that multipath will be a big issue and might corrupt the measurement. The loco positioning system UWB radio solves that by sending very short impulse that uses a Ultra Wide Bandwidth.

As for the log question, logs currently only supports periodic logs, up to 100Hz by default, 1KHz is the technical limit. The way to send asynchronous event is to send packet for example using the app channel. about 1000 packets per seconds is the limit for communication with Crazyradio.

**Reply by shossains** [reply]
Thank you so much!

I guess I will scrap ranging for now. I will focus more in the direction of decentralized formation control by sending positions P2P (at a much lower reasonable frequency).

If anything comes up I'll be sure to open up a new discussion
